core/src/worker/mod.rs

```diff
@@ -69,6 +72,10 @@ pub struct Worker {
     /// Will be called at the end of each activation completion
     #[allow(clippy::type_complexity)] // Sorry clippy, there's no simple way to re-use here.
     post_activate_hook: Option<Box<dyn Fn(&Self, &str, usize) + Send + Sync>>,
+    /// Set when non-local activities are complete and should stop being polled
+    non_local_activities_complete: Arc<AtomicBool>,
+    /// Set when local activities are complete and should stop being polled
+    local_activities_complete: Arc<AtomicBool>,
```

Comment:
Ended up going with these atomics since I didn't want to do another refactor.
Can revisit in a follow up PR.

Reply:
Local and non-local activities should be aggregated into a single stream instead.

Reply:
I just didn't want this to drag on longer than it already is.
I agree a stream is the right approach.